Non-deterministic behavior in featurization

**Describe the bug**
When working with large (~7k docs) corpus of hardware datasheets, extracting multiple relations, we expect that the features for each candidate would be deterministic between each run. Even more so if we have `parallelism=1` set in the Featurizer. However, we find that there can be small (e.g., < 5) differences between feature tables, resulting in slightly different sparse matrices, and thus, slightly different results.

**To Reproduce**
Running on the HACK transistor dataset will reproduce the error. However, it will take a long time, and we haven't been able to get a very minimal example that reproduces the error yet. Attached are two feature table dumps between two different runs with `parallelism=1`. Note that there is only a single difference on line `65454`.

[feature_table.tar.gz](https://github.com/HazyResearch/fonduer/files/4589655/feature_table.tar.gz)

Note that it isn't always one difference, and the difference is not deterministic. The different attached is just an example.

**Expected behavior**
We would expect that these feature tables are identical between runs.

**Error Logs/Screenshots**
For convenience, here is the differing line in screenshot form
![image](https://user-images.githubusercontent.com/7573542/81232657-1e131200-8faa-11ea-9d84-8a735b693166.png)

**Additional context**
If the issue is in the UDF implementation, this might affect the Labeler in addition to the Featurizer, since they share a lot of the UDF code.

I'm working on #439 and currently the tests fail https://github.com/HazyResearch/fonduer/pull/439/checks?check_run_id=759815347 at assertions on the number of feature keys like below.
Mysteriously, the tests fail only at the number of feature keys while other assertions pass.

```
...
        assert num_features == len(train_cands[0]) + len(train_cands[1])
        num_feature_keys = session.query(FeatureKey).count()
>       assert num_feature_keys == 4577
E       assert 4621 == 4577

tests/e2e/test_e2e.py:235: AssertionError

...
        # Update features
        featurizer.update(new_docs, parallelism=PARALLEL)
        assert session.query(Feature).count() == len(train_cands[0])
        num_feature_keys = session.query(FeatureKey).count()
>       assert num_feature_keys == 2526
E       assert 2544 == 2526

tests/e2e/test_incremental.py:192: AssertionError
```

It seems to be reproducible as the same commit fails on my forked repo too (https://github.com/HiromuHota/fonduer/runs/759811429?check_suite_focus=true).

One more thing: the same GitHub Actions job (https://github.com/HazyResearch/fonduer/pull/439/checks?check_run_id=759815303), but a different error on mac (3.7)

```
        # Test that FeatureKey is properly reset
        featurizer.apply(split=1, train=True, parallelism=PARALLEL)
>       assert session.query(Feature).count() == 214
E       assert 213 == 214
E        +  where 213 = <bound method Query.count of <sqlalchemy.orm.query.Query object at 0x138f72c50>>()
E        +    where <bound method Query.count of <sqlalchemy.orm.query.Query object at 0x138f72c50>> = <sqlalchemy.orm.query.Query object at 0x138f72c50>.count
E        +      where <sqlalchemy.orm.query.Query object at 0x138f72c50> = <bound method Session.query of <sqlalchemy.orm.session.Session object at 0x1355437b8>>(Feature)
E        +        where <bound method Session.query of <sqlalchemy.orm.session.Session object at 0x1355437b8>> = <sqlalchemy.orm.session.Session object at 0x1355437b8>.query

tests/e2e/test_e2e.py:188: AssertionError
```

Two differences:
- test_e2e.py failed both Mac (3.6) and Linux (3.7), but it failed on the number of features on Mac.
- test_incremental.py passed on Mac.

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -43,6 +43,9 @@ Fixed
 * `@HiromuHota`_: Fix the order of args to Bbox.
   (`#443 <https://github.com/HazyResearch/fonduer/issues/443>`_)
   (`#444 <https://github.com/HazyResearch/fonduer/pull/444>`_)
+* `@HiromuHota`_: Fix the non-deterministic behavior in VisualLinker.
+  (`#412 <https://github.com/HazyResearch/fonduer/issues/412>`_)
+  (`#458 <https://github.com/HazyResearch/fonduer/pull/458>`_)
 
 0.8.2_ - 2020-04-28
 -------------------

--- a/src/fonduer/parser/visual_linker.py
+++ b/src/fonduer/parser/visual_linker.py
@@ -6,6 +6,7 @@ import shutil
 import subprocess
 from builtins import object, range, zip
 from collections import OrderedDict, defaultdict
+from operator import attrgetter
 from typing import DefaultDict, Dict, Iterator, List, Optional, Tuple
 
 import numpy as np
@@ -63,7 +64,8 @@ class VisualLinker(object):
             override the one used in initialization, if provided.
         :return: A generator of ``Sentence``.
         """
-        self.sentences = sentences
+        # sentences should be sorted as their order is not deterministic.
+        self.sentences = sorted(sentences, key=attrgetter("position"))
         self.pdf_file = (
             pdf_path if os.path.isfile(pdf_path) else pdf_path + document_name + ".pdf"
         )

--- a/tests/parser/test_visual_linker.py
+++ b/tests/parser/test_visual_linker.py
@@ -1,0 +1,54 @@
+"""Fonduer visual_linker unit tests."""
+import random
+from operator import attrgetter
+
+from fonduer.parser.preprocessors import HTMLDocPreprocessor
+from fonduer.parser.visual_linker import VisualLinker
+from tests.parser.test_parser import get_parser_udf
+
+
+def test_visual_linker_not_affected_by_order_of_sentences():
+    """Test if visual_linker result is not affected by the order of sentences."""
+    docs_path = "tests/data/html/2N6427.html"
+    pdf_path = "tests/data/pdf/2N6427.pdf"
+
+    # Initialize preprocessor, parser, visual_linker.
+    # Note that parser is initialized with `visual=False` and that visual_linker
+    # will be used to attach "visual" information to sentences after parsing.
+    preprocessor = HTMLDocPreprocessor(docs_path)
+    parser_udf = get_parser_udf(
+        structural=True, lingual=False, tabular=True, visual=False
+    )
+    visual_linker = VisualLinker(pdf_path=pdf_path)
+
+    doc = parser_udf.apply(next(preprocessor.__iter__()))
+    # Sort sentences by sentence.position
+    doc.sentences = sorted(doc.sentences, key=attrgetter("position"))
+    sentences0 = [
+        sent for sent in visual_linker.link(doc.name, doc.sentences, pdf_path)
+    ]
+    # Sort again in case visual_linker.link changes the order
+    sentences0 = sorted(sentences0, key=attrgetter("position"))
+
+    doc = parser_udf.apply(next(preprocessor.__iter__()))
+    # Shuffle
+    random.shuffle(doc.sentences)
+    sentences1 = [
+        sent for sent in visual_linker.link(doc.name, doc.sentences, pdf_path)
+    ]
+    # Sort sentences by sentence.position
+    sentences1 = sorted(sentences1, key=attrgetter("position"))
+
+    # This should hold as both sentences are sorted by their position
+    assert all(
+        [
+            sent0.position == sent1.position
+            for (sent0, sent1) in zip(sentences0, sentences1)
+        ]
+    )
+
+    # The following assertion should hold if the visual_linker result is not affected
+    # by the order of sentences.
+    assert all(
+        [sent0.left == sent1.left for (sent0, sent1) in zip(sentences0, sentences1)]
+    )

